homeassistant/components/recorder/migration.py

```diff
@@ -41,20 +41,39 @@ def _apply_update(engine, new_version):
     from sqlalchemy import Table
     from . import models
 
-    if new_version == 1:
-        def create_index(table_name, column_name):
-            """Create an index for the specified table and column."""
-            table = Table(table_name, models.Base.metadata)
-            name = "_".join(("ix", table_name, column_name))
-            # Look up the index object that was created from the models
-            index = next(idx for idx in table.indexes if idx.name == name)
-            _LOGGER.debug("Creating index for table %s column %s",
-                          table_name, column_name)
-            index.create(engine)
-            _LOGGER.debug("Index creation done for table %s column %s",
-                          table_name, column_name)
+    def create_index(table_name, column_name):
+        """Create an index for the specified table and column."""
+        table = Table(table_name, models.Base.metadata)
+        name = "_".join(("ix", table_name, column_name))
```

Comment:
So what happens when we load an old database with new models objects, is that the actual Index sqlalchemy metadata objects get initialized with all the parameters we used in models.py, but the index isn't created in the underlying database engine. These indexes are present in the `table.indexes` list.

[next](https://docs.python.org/3/library/functions.html#next) is a python built-in that returns the next item from an iterator. Since we create a new iterator by searching through table.indexes for idx.name == name, there's only going to be one index in that iterator. What that means is that we're getting the metadata object for the index that was created by models.py. Once we have the reference all we have to do is call index.create.